lib/openhab/dsl/debouncer.rb

```diff
@@ -0,0 +1,241 @@
+
+      #
+      # Debounce calls to the given block.
+      #
+      # This method is meant to be called repeatedly with the same given block.
+      # However, if no block is given, it will call and debounce the previously given block
+      #
+      # @yield Block to be debounced
+      #
+      # @return [void]
+      #
+      # @example Basic trailing edge debouncing
+      #   debouncer = Debouncer.new(for: 1.minute)
+      #   (1..100).each do
+      #     debouncer.call { logger.info "I won't log more often than once a minute" }
+      #     sleep 20 # call the debouncer every 20 seconds
+      #   end
+      #
+      # @example Call the previous debounced block
+      #   debouncer = Debouncer.new(for: 1.minute)
+      #   debouncer.call { logger.info "Hello. It is #{Time.now}" } # First call to debounce
+      #
+      #   after(20.seconds) do |timer|
+      #     debouncer.call # Call the original block above
+      #     timer.reschedule unless timer.cancelled?
+      #   end
+      #
+      def call(&block)
+        @block = block if block
+        raise ArgumentError, "No block has been provided" unless @block
+
+        return call! unless @interval # passthrough mode, no debouncing when @interval is nil
+
+        now = ZonedDateTime.now
+
+        if leading?
+          leading_edge_debounce(now)
+        else
+          trailing_edge_debounce(now)
+        end
+        @mutex.synchronize { @last_timestamp = now }
+      end
+
+      #
+      # Execute the latest block passed to the {#debounce} call regardless of any debounce settings.
+      #
+      # @return [Object] The return value of the block
+      #
+      def call!
+        @block.call
+      end
+
+      #
+      # Resets the debounce period and cancels any outstanding block executions of a trailing edge debouncer.
+      #
+      # - A leading edge debouncer will execute its block on the next call and start a new debounce period.
+      # - A trailing edge debouncer will reset its debounce timer and the next call will become the start
+      #   of a new debounce period.
+      #
+      # @return [Boolean] True if a pending execution was cancelled.
+      #
+      def reset
+        @mutex.synchronize do
+          if leading?
+            @last_timestamp = @leading_timestamp = ZonedDateTime.of(1900, 1, 1, 0, 0, 0, 0,
+                                                                    ZoneId.system_default)
+          end
+          @debounce_timer&.cancel
+        end
+      end
+
+      #
+      # Immediately executes any outstanding event of a trailing edge debounce.
+      # The next call will start a new period.
+      #
+      # It has no effect on a leading edge debouncer - use {#reset} instead.
+      #
+      # @return [Boolean] True if an existing debounce timer was rescheduled to run immediately.
+      #   False if there were no outstanding executions.
+      #
+      def flush
+        @mutex.synchronize { call! && true if @debounce_timer&.cancel }
```

Comment:
what if `call!` (the user-provided block) returns `nil` or `false`? his should probably just be written as a multi-statement non-modifier `if`

Reply:
Good catch.